### DISCUSSION TITLE
Metaprogramming

### DISCUSSION THREAD

**Discussion by DaleStrickler** [dotnet/roslyn#51713] [Ideas]
I have been a fan of [metaprogramming](https://en.wikipedia.org/wiki/Metaprogramming) for decades.

I am relatively new to the C# environment and I am only using it via Unity and JetBrains Rider on Mac.

I want to write C# code that generates C# code based on some rules, directives, and existing code analysis.

I am looking for any help I can get to find the relevant parts of the compiler source that would help me find relevant constructs within C# solutions.  As a very simple example, how could I use the Roslyn code to help me create a code generator that added a log message to all constructors (static, multi-signature, etc.) within a C# solution?

I just cloned the Roslyn repo tonight.  I am just now starting to sort through it.

My fantasy is to have a 'regex'-like system that understands a higher level of C# abstraction. Allowing a match pattern that would represent a high-level concept.  Something like "all public class attributes containing the string 'foo'". Then allow replacements like "replace the simple public attributes with private field references that have a public get and a private set."

I am very new to the C# ecosystem.  Maybe there are tools/techniques for just this sort of abstract code concept identification.  I spent some time searching for "Metaprogramming" and other keywords I could think of in the Roslyn project and I have found nothing yet.  I will keep looking for ideas, as well as reading the Roslyn solution.  Any shortcuts towards my metaprogramming fantasies that any of you could offer would be great.  Thanks!

**Comment by svick** [comment]
The closest you can get to what you're looking for with Microsoft-supported offerings are [source generators](https://devblogs.microsoft.com/dotnet/new-c-source-generator-samples/) (though they only let you generate new code, so you can't use them to add a statement to a constructor) and IDE refactorings (but those modify the code you see).

To actually get what you want, you'll need to look at 3rd party solutions. There are two common tools that use IL rewriting for this: the open-source [Fody](https://github.com/Fody/Fody) and the commercial [PostSharp](https://www.postsharp.net/) (both already have plugins for this kind of logging/tracing).

If you're looking for something that's based on manipulating C# code and using Roslyn APIs, then the only tool I'm aware of is [PostSharp Caravela](https://github.com/postsharp/Caravela), which is in early preview. (Disclaimer: I have worked on that project.)

**Reply by DaleStrickler** [reply]
Great leads.  I look forward to following up on these.

I have made some minimal progress by getting really good with regex substitutions but it gets ugly and makes a lot of assumptions about code layout.
